Java3practise4/J2_Stream201126/chat-server/src/ru/geekbrains/java_two/chat/server/gui/ServerGUI.java

```diff
@@ -0,0 +1,89 @@
+package ru.geekbrains.java_two.chat.server.gui;
+
+import ru.geekbrains.java_two.chat.server.core.ChatServer;
+import ru.geekbrains.java_two.chat.server.core.ChatServerListener;
+
+import javax.swing.*;
+import java.awt.*;
+import java.awt.event.ActionEvent;
+import java.awt.event.ActionListener;
+
+public class ServerGUI extends JFrame implements ActionListener,
+        Thread.UncaughtExceptionHandler, ChatServerListener {
+
+    private static final int POS_X = 800;
+    private static final int POS_Y = 200;
+    private static final int WIDTH = 600;
+    private static final int HEIGHT = 300;
+
+    private final ChatServer chatServer = new ChatServer(this);
+    private final JButton btnStart = new JButton("Start");
+    private final JButton btnStop = new JButton("Stop");
+    private final JPanel panelTop = new JPanel(new GridLayout(1, 2));
+    private final JTextArea log = new JTextArea();
+
+    public static void main(String[] args) {
+        SwingUtilities.invokeLater(new Runnable() { // Event Dispatching Thread
+            @Override
+            public void run() {
+                new ServerGUI(); // Event Queue
+            }
+        });
+    }
+
+    private ServerGUI() {
+        Thread.setDefaultUncaughtExceptionHandler(this);
+        setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
+        setBounds(POS_X, POS_Y, WIDTH, HEIGHT);
+        setResizable(false);
+        setTitle("Chat server");
+        setAlwaysOnTop(true);
+        log.setEditable(false);
+        log.setLineWrap(true);
+        JScrollPane scrollLog = new JScrollPane(log);
+        btnStop.addActionListener(this);
+        btnStart.addActionListener(this);
+
+        panelTop.add(btnStart);
+        panelTop.add(btnStop);
+        add(panelTop, BorderLayout.NORTH);
+        add(scrollLog, BorderLayout.CENTER);
+        setVisible(true);
+    }
+
+    @Override
+    public void actionPerformed(ActionEvent e) {
+        Object src = e.getSource();
+        if (src == btnStart) {
+            chatServer.start(8189);
+        } else if (src == btnStop) {
+            //throw new RuntimeException("Hello from EDT!");
+            chatServer.stop();
+        } else {
+            throw new RuntimeException("Undefined source: " + src);
+        }
+    }
+
+    @Override
+    public void uncaughtException(Thread t, Throwable e) {
+        e.printStackTrace();
+        StackTraceElement[] ste = e.getStackTrace();
+        String msg = String.format("Exception in thread \"%s\" %s: %s\n\tat %s",
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#fs-format-string-should-use-n-rather-than-n-va-format-string-uses-newline).
This format string includes a newline character (\n). In format strings, it is generally  preferable to use %n, which will produce the platform-specific line separator.